Greedy/GaleShapley.js

```diff
@@ -0,0 +1,78 @@
+/**
+ * The Gale-Shapley algorithm is used to find a stable matching between two sets
+ * of equal size (e.g., donors and recipients) where no two elements prefer each
+ * other over their current partners. It ensures a stable matching by proposing
+ * from one side to the other until all are matched.
+ *
+ * Complexity:
+ *    Worst-case performance O(n^2)
+ *    Best-case performance O(n^2)
+ *
+ * Reference:
+ *    https://en.wikipedia.org/wiki/Stable_marriage_problem
+ *    https://www.youtube.com/watch?v=Qcv1IqHWAzg (Numberphile)
+ *
+ */
+
+/**
+ *
+ * @param {number[][]} donorPref Preferences of donors, where each donor has an ordered list of recipients.
+ * @param {number[][]} recipientPref Preferences of recipients, where each recipient has an ordered list of donors.
+ * @returns {number[]} Array where the index is the donor, and the value at the index is the matched recipient.
+ *
+ * @example
+ * const donorPref = [[0, 1, 3, 2], [0, 2, 3, 1], [1, 0, 2, 3], [0, 3, 1, 2]];
+ * const recipientPref = [[3, 1, 2, 0], [3, 1, 0, 2], [0, 3, 1, 2], [1, 0, 3, 2]];
+ * stableMatching(donorPref, recipientPref); // Output: [1, 2, 3, 0]
+ */
+function stableMatching(donorPref, recipientPref) {
+  // Initialize the number of donors and create a list of unmatched donors
+  let n = donorPref.length
+  let unmatchedDonors = Array.from({ length: n }, (_, i) => i)
+
+  // Records of which recipient each donor is paired with and vice versa
+  let donorRecord = Array(n).fill(-1) // Donor to recipient
```

Comment:
I'd compute `donorRecord` at the end (or not at all even and instead just return `recRecord` and document that because the two are equivalent). That way you save yourself having to remember to update it during the algorithm, and the constant factor of the $O(n^2)$ many operations is better.